isobus/src/isobus_task_controller_client.cpp

```diff
@@ -623,14 +744,170 @@ namespace isobus
+						{
+							ddopLocalizationLabel[i] = (userSuppliedBinaryDDOP[currentByteIndex + i]); // Read the localization label
+						}
+						break;
+					}
+				}
+			}
+			break;
+
+			case DDOPUploadType::UserProvidedVector:
+			{
+				// Searching for "DVC"
+				while (currentByteIndex < (generatedBinaryDDOP.size() - DEVICE_TABLE_ID.size()))
+				{
+					if ((DEVICE_TABLE_ID[0] == generatedBinaryDDOP.at(currentByteIndex)) &&
+					    (DEVICE_TABLE_ID[1] == generatedBinaryDDOP.at(currentByteIndex + 1)) &&
+					    (DEVICE_TABLE_ID[2] == generatedBinaryDDOP.at(currentByteIndex + 2)))
+					{
+						// We have to do a lot of error checking on the DDOP length
+						// This is because we don't control the content of this DDOP, and have no
+						// assurances that the schema is even valid
+
+						assert((currentByteIndex + DESIGNATOR_BYTE_OFFSET) < generatedBinaryDDOP.size()); // Not enough bytes to read the designator length
+						currentByteIndex += DESIGNATOR_BYTE_OFFSET; // Skip to the next variable length part of the object
+
+						const std::uint32_t DESIGNATOR_LENGTH = generatedBinaryDDOP.at(currentByteIndex); // "N", See Table A.1
+						assert(currentByteIndex + DESIGNATOR_LENGTH < generatedBinaryDDOP.size()); // Not enough bytes in your DDOP!
+						currentByteIndex += DESIGNATOR_LENGTH + 1;
+
+						const std::uint32_t SOFTWARE_VERSION_LENGTH = generatedBinaryDDOP.at(currentByteIndex); // "M", See Table A.1
+						assert(currentByteIndex + SOFTWARE_VERSION_LENGTH + CLIENT_NAME_LENGTH < generatedBinaryDDOP.size()); // Not enough bytes in your DDOP!
+						currentByteIndex += SOFTWARE_VERSION_LENGTH + CLIENT_NAME_LENGTH + 1;
+
+						const std::uint32_t SERIAL_NUMBER_LENGTH = generatedBinaryDDOP.at(currentByteIndex); // "O", See Table A.1
+						assert(currentByteIndex + SERIAL_NUMBER_LENGTH < generatedBinaryDDOP.size()); // Not enough bytes in your DDOP!
+						currentByteIndex += SERIAL_NUMBER_LENGTH + 1;
+
+						assert(currentByteIndex + task_controller_object::DeviceObject::MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH < generatedBinaryDDOP.size()); // // Not enough bytes in your DDOP!
+						for (std::uint_fast8_t i = 0; i < task_controller_object::DeviceObject::MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH; i++)
+						{
+							ddopStructureLabel.push_back(generatedBinaryDDOP.at(currentByteIndex + i)); // Read the structure label
+						}
+
+						currentByteIndex += task_controller_object::DeviceObject::MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH;
+						assert(currentByteIndex + task_controller_object::DeviceObject::MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH < generatedBinaryDDOP.size()); // // Not enough bytes in your DDOP!
+						for (std::uint_fast8_t i = 0; i < task_controller_object::DeviceObject::MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH; i++)
+						{
+							ddopLocalizationLabel[i] = (generatedBinaryDDOP.at(currentByteIndex + i)); // Read the localization label
+						}
+						break;
+					}
+				}
```

Comment:
We might be able to shorten this by using a lambda function? Afaik it will only run once on startup, so it won't have that much of an performance impact if any.

Something like
```cpp
	auto getDDOPSize = [&ddopUploadMode, &userSuppliedBinaryDDOPSize_bytes, &generatedBinaryDDOP]() {
		if (ddopUploadMode == DDOPUploadMode::UserProvidedBinaryPointer)
		{
			return userSuppliedBinaryDDOPSize_bytes;
		}
		else
		{
			return generatedBinaryDDOP.size();
		}
	};

	auto getDDOPByteAt = [&ddopUploadMode, &userSuppliedBinaryDDOPSize_bytes, &generatedBinaryDDOP](std::size_t index) {
		if (ddopUploadMode == DDOPUploadMode::UserProvidedBinaryPointer)
		{
			return userSuppliedBinaryDDOP[index];
		}
		else
		{
			return generatedBinaryDDOP.at(index);
		}
	};
```
might work?

Otherwise we also have to keep the code in the two cases synced, which I'm not really good at hahah

Reply:
It does feel like a lot of duplication, doesn't it? I can try your solution!